### DISCUSSION TITLE
MACR usage?

### DISCUSSION THREAD

**Discussion by cprima** [gopro/labs#331] [Q&A] [answered]
The release notes for 2.10.70 mention

> Added crude super of macro (near focus) lenses, with e.g. oMMACR=-0.1,0.2,… (add the distortion prescription for the added lens.)

I am intrigued! But on https://gopro.github.io/labs/control/extensions/ there is no mention of MACR. What does it do?

**Comment by dnewman-gpsw** [comment, marked as answer]
It is from an incomplete hackathon project, but it works enough to share.  The idea: when you add a macro lens, the overall lens prescription changes, causing Hypersmooth to have wobbles on the edges.  A lens model is written a polynomial  world_angle = ar + br^2 + cr^3 + ...   As the internal lens model is not shareable (IP), the MACR is a lens modifier so is you try oMMACR=-0.1,0.2 then camera will use world_angle = (a-0.1)r + (b+0.2)r^2 + cr^3 + ... 

You problem is you will need to work out the prescription correction. This is example is not correct, but it shows the MACR is modifying the curvature.

![image](https://user-images.githubusercontent.com/10856084/235759316-c13249b6-c425-45a3-ae15-314a8eb3720a.png)

**Reply by cprima** [reply]
Thanks for the info, I have a macro lens adapter in my Amazon cart, maybe get it next month.
https://www.amazon.de/gp/product/B09MW3NF54/ , you seem to reference this one https://www.bhphotovideo.com/c/product/1648747-REG/flip_filters_ff_mm9_flip9_15_macromate_mini.html

Anyhow, will give this a try once I have a macro lens, thank you for the explanation plus pictures!

**Reply by IRCGraphic** [reply]
Could this barrel distortion correction model be applied to a circular fisheye like the Max lens mod?
(I'm struggling to do post stabilisation with the Max lens mod even with the BERS=1 code)

**Reply by dnewman-gpsw** [reply]
No idea, that would take some effort.

**Reply by IRCGraphic** [reply]
I eyeballed inverting the signs of those two values... to my surprise it was able to unwrap a full circular fisheye but it created little graphic glitches and wrapping distortions around the periphery.
According to you if a precise model is tailored it could be used with stabilization calibrated with GoPro's internal electronic rolling shutter correction?
I have .json and lightroom's .lcp lens profiles of the Max lens for rectilinear unwrapping, those calibration "distortion_coeffs" are any useful for your polynomial world_angle model?
I've tried Max lens mod post stabilization with BERS=1 and even with a pretty close eyeballed rolling shutter correction of 25 ms for some reason it's still super jelly compared to other post-stabilised action cameras I hacked with circular fisheyes.
It would be really advantageous if Max lens could:
a) be used with its full FOV potential, it doesn't need more than a standard stabilization IMO the super-wide angles side effect is that are more pleasant to movement 
b) use the in-camera stabilization, post stabilisation is heavy computing (for me) specially with those 5K or 10-bit videos
